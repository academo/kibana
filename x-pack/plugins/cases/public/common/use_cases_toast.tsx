/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { EuiButtonEmpty, EuiText } from '@elastic/eui';
import React from 'react';
import styled from 'styled-components';
import { toMountPoint } from '../../../../../src/plugins/kibana_react/public';
import { Case } from '../../common';
import { useToasts } from '../common/lib/kibana';
import { useCaseViewNavigation } from '../common/navigation';
import { CASE_SUCCESS_SYNC_TEXT, CASE_SUCCESS_TOAST, VIEW_CASE } from './translations';

const LINE_CLAMP = 3;
const Title = styled.span`
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: ${LINE_CLAMP};
  -webkit-box-orient: vertical;
  overflow: hidden;
`;
const EuiTextStyled = styled(EuiText)`
  ${({ theme }) => `
    margin-bottom: ${theme.eui?.paddingSizes?.s ?? 8}px;
  `}
`;

export const useCasesToast = () => {
  const { navigateToCaseView } = useCaseViewNavigation();

  const toasts = useToasts();

  return {
    showSuccessAttach: (
      theCase: Case,
      { title, content }: { title?: string; content?: string } = {}
    ) => {
      const onViewCaseClick = () => {
        navigateToCaseView({
          detailName: theCase.id,
        });
      };
      return toasts.addSuccess({
        color: 'success',
        iconType: 'check',
        title: toMountPoint(<Title>{title ? title : CASE_SUCCESS_TOAST(theCase.title)}</Title>),
        text: toMountPoint(
          <CaseToastSuccessContent
            content={content}
            syncAlerts={theCase.settings.syncAlerts}
            onViewCaseClick={onViewCaseClick}
          />
        ),
      });
    },
  };
};
export const CaseToastSuccessContent = ({
  syncAlerts,
  onViewCaseClick,
  content,
}: {
  syncAlerts: boolean;
  onViewCaseClick: () => void;
  content?: string;
}) => {
  let toastContent: string | undefined;
  if (content !== undefined) {
    toastContent = content;
  } else if (syncAlerts) {
    toastContent = CASE_SUCCESS_SYNC_TEXT;
  }

  return (
    <>
      {toastContent !== undefined ? (
        <EuiTextStyled size="s" data-test-subj="toaster-content-sync-text">
          {toastContent}
        </EuiTextStyled>
      ) : null}
      <EuiButtonEmpty
        size="xs"
        flush="left"
        onClick={onViewCaseClick}
        data-test-subj="toaster-content-case-view-link"
      >
        {VIEW_CASE}
      </EuiButtonEmpty>
    </>
  );
};
CaseToastSuccessContent.displayName = 'CaseToastSuccessContent';
