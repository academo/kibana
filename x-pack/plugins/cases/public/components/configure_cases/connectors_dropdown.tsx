/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React, { ReactNode, Suspense, useCallback, useEffect, useMemo, useState } from 'react';
import {
  EuiComboBox,
  EuiComboBoxOptionOption,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiIconTip,
} from '@elastic/eui';
import styled from 'styled-components';

import { ActionConnector } from '../../containers/configure/types';
import * as i18n from './translations';
import { useKibana } from '../../common/lib/kibana';
import { getConnectorIcon, isDeprecatedConnector } from '../utils';
import { euiStyled } from '../../../../../../src/plugins/kibana_react/common';

type ConnectorEuiComboBoxOptionOption<T> = EuiComboBoxOptionOption<T> & {
  rendered: ReactNode;
};

export interface Props {
  connectors: ActionConnector[];
  disabled: boolean;
  isLoading: boolean;
  onChange: (id: string) => void;
  selectedConnector: string;
  appendAddConnectorButton?: boolean;
}

const ICON_SIZE = 'm';

const EuiIconExtended = styled(EuiIcon)`
  margin-right: 13px;
  margin-bottom: 0 !important;
`;

const FlexItemIconContainer = styled(EuiFlexItem)`
  svg {
    margin-right: 13px;
    margin-bottom: 0 !important;
    width: 20px;
  }
`;

const noConnectorOption = {
  value: 'none',
  label: i18n.NO_CONNECTOR,
  'data-test-subj': 'dropdown-connector-no-connector',
  rendered: (
    <EuiFlexGroup gutterSize="none" alignItems="center" responsive={false}>
      <EuiFlexItem grow={false}>
        <EuiIconExtended type="minusInCircle" size={ICON_SIZE} />
      </EuiFlexItem>
      <EuiFlexItem>
        <span data-test-subj={`dropdown-connector-no-connector`}>{i18n.NO_CONNECTOR}</span>
      </EuiFlexItem>
    </EuiFlexGroup>
  ),
};

const addNewConnector = {
  value: 'add-connector',
  label: i18n.ADD_NEW_CONNECTOR,
  'data-test-subj': 'dropdown-connector-add-connector',
  render: (
    <span className="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushLeft">
      {i18n.ADD_NEW_CONNECTOR}
    </span>
  ),
};

const StyledEuiIconTip = euiStyled(EuiIconTip)`
  margin-left: ${({ theme }) => theme.eui.euiSizeS}
  margin-bottom: 0 !important;
`;

const ConnectorsDropdownComponent: React.FC<Props> = ({
  connectors,
  disabled,
  isLoading,
  onChange,
  selectedConnector,
  appendAddConnectorButton = false,
}) => {
  const { triggersActionsUi } = useKibana().services;

  const getConnectorIconSuspended = useCallback(
    (connector: ActionConnector) => {
      const LazyIcon = getConnectorIcon(triggersActionsUi, connector.actionTypeId);
      return (
        <Suspense fallback="<span></span>">
          <LazyIcon />
        </Suspense>
      );
    },
    [triggersActionsUi]
  );

  const connectorsAsDropdownOptions: Array<ConnectorEuiComboBoxOptionOption<string>> =
    useMemo(() => {
      const connectorsFormatted = connectors.reduce(
        (acc, connector) => {
          return [
            ...acc,
            {
              value: connector.id,
              label: connector.name,
              'data-test-subj': `dropdown-connector-${connector.id}`,
              rendered: (
                <EuiFlexGroup gutterSize="s" alignItems="center" responsive={false}>
                  <FlexItemIconContainer grow={false}>
                    <EuiIconExtended
                      type={() => getConnectorIconSuspended(connector)}
                      size={ICON_SIZE}
                    />
                  </FlexItemIconContainer>
                  <EuiFlexItem grow={false}>
                    <span>
                      {connector.name}
                      {isDeprecatedConnector(connector) && ` (${i18n.DEPRECATED_TOOLTIP_TEXT})`}
                    </span>
                  </EuiFlexItem>
                  {isDeprecatedConnector(connector) && (
                    <EuiFlexItem grow={false}>
                      <StyledEuiIconTip
                        aria-label={i18n.DEPRECATED_TOOLTIP_CONTENT}
                        size={ICON_SIZE}
                        type="alert"
                        color="warning"
                        content={i18n.DEPRECATED_TOOLTIP_CONTENT}
                      />
                    </EuiFlexItem>
                  )}
                </EuiFlexGroup>
              ),
            },
          ];
        },
        [noConnectorOption]
      );

      if (appendAddConnectorButton) {
        return [...connectorsFormatted, addNewConnector];
      }

      return connectorsFormatted;
    }, [appendAddConnectorButton, connectors, getConnectorIconSuspended]);

  const [selected, setSelected] = useState<Array<EuiComboBoxOptionOption<string>>>([
    noConnectorOption,
  ]);

  // handle the existing external selection
  useEffect(() => {
    if (selected?.length > 0 && selectedConnector !== selected[0].value) {
      const selectedOption = connectorsAsDropdownOptions.find(
        (connector) => connector.value === selectedConnector
      );
      if (selectedOption) {
        setSelected([selectedOption]);
      }
    }
  }, [connectorsAsDropdownOptions, selected, selectedConnector]);

  const innerOnChange = (option: Array<EuiComboBoxOptionOption<string>>) => {
    setSelected(option);
    if (option?.length > 0 && option[0].value) {
      onChange(option[0].value);
    }
  };

  // if the user leaves the combobox with an empty input
  // it will reset the one selected before or to no connector
  // if none was selected before
  const handleOnBlur = () => {
    if (selected?.length === 0) {
      const previousSelection = connectorsAsDropdownOptions.find(
        (connector) => connector.value === selectedConnector
      );
      if (previousSelection) {
        innerOnChange([previousSelection]);
      } else {
        innerOnChange([noConnectorOption]);
      }
    }
  };

  const renderOption = (option: EuiComboBoxOptionOption<string>, _searchString: string) => {
    return (option as ConnectorEuiComboBoxOptionOption<string>).rendered;
  };

  return (
    <EuiComboBox
      aria-label={i18n.INCIDENT_MANAGEMENT_SYSTEM_LABEL}
      data-test-subj="dropdown-connectors"
      isDisabled={disabled}
      renderOption={renderOption}
      fullWidth
      isLoading={isLoading}
      onChange={innerOnChange}
      singleSelection={{ asPlainText: true }}
      options={connectorsAsDropdownOptions}
      selectedOptions={selected}
      onBlur={handleOnBlur}
    />
  );
};

ConnectorsDropdownComponent.displayName = 'ConnectorsDropdown';

export const ConnectorsDropdown = React.memo(ConnectorsDropdownComponent);
