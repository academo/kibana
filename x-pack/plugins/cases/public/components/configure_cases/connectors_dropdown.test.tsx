/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { EuiComboBox } from '@elastic/eui';
import { act, fireEvent, render, screen } from '@testing-library/react';

import { ConnectorsDropdown, Props } from './connectors_dropdown';
import { TestProviders } from '../../common/mock';
import { connectors } from './__mock__';
import { useKibana } from '../../common/lib/kibana';
import { registerConnectorsToMockActionRegistry } from '../../common/mock/register_connectors';
import userEvent from '@testing-library/user-event';

jest.mock('../../common/lib/kibana');
const useKibanaMock = useKibana as jest.Mocked<typeof useKibana>;

describe('ConnectorsDropdown', () => {
  let wrapper: ReactWrapper;
  const props: Props = {
    disabled: false,
    connectors,
    isLoading: false,
    onChange: jest.fn(),
    selectedConnector: 'none',
  };

  const actionTypeRegistry = useKibanaMock().services.triggersActionsUi.actionTypeRegistry;

  beforeAll(() => {
    registerConnectorsToMockActionRegistry(actionTypeRegistry, connectors);
    wrapper = mount(<ConnectorsDropdown {...props} />, { wrappingComponent: TestProviders });
  });

  it('renders', () => {
    expect(wrapper.find('[data-test-subj="dropdown-connectors"]').first().exists()).toBe(true);
  });

  it('formats the connectors correctly', () => {
    const selectProps = wrapper.find(EuiComboBox).props();

    expect(selectProps.options).toMatchSnapshot();
  });

  it('disables the dropdown', () => {
    const newWrapper = mount(<ConnectorsDropdown {...props} disabled={true} />, {
      wrappingComponent: TestProviders,
    });

    expect(
      newWrapper.find('[data-test-subj="dropdown-connectors"]').first().prop('isDisabled')
    ).toEqual(true);
  });

  it('shows loading correctly', () => {
    const newWrapper = mount(<ConnectorsDropdown {...props} isLoading={true} />, {
      wrappingComponent: TestProviders,
    });

    expect(
      newWrapper.find('[data-test-subj="dropdown-connectors"]').first().prop('isLoading')
    ).toEqual(true);
  });

  it('selects the previously selected connector correctly', () => {
    const newWrapper = mount(<ConnectorsDropdown {...props} selectedConnector={'servicenow-1'} />, {
      wrappingComponent: TestProviders,
    });

    expect(newWrapper.find('[data-test-subj="dropdown-connectors"]').first().text()).toBe(
      'My Connector'
    );
  });

  it('does not throw when accessing the icon if the connector type is not registered', () => {
    expect(() =>
      mount(
        <ConnectorsDropdown
          {...props}
          connectors={[
            {
              id: 'none',
              actionTypeId: '.none',
              name: 'None',
              config: {},
              isPreconfigured: false,
            },
          ]}
        />,
        {
          wrappingComponent: TestProviders,
        }
      )
    ).not.toThrowError();
  });

  it('shows the deprecated tooltip when the connector is deprecated', () => {
    render(<ConnectorsDropdown {...props} selectedConnector="servicenow-uses-table-api" />, {
      wrapper: ({ children }) => <TestProviders>{children}</TestProviders>,
    });

    userEvent.click(screen.getByTestId('comboBoxToggleListButton'));

    const tooltips = screen.getAllByLabelText(
      'This connector is deprecated. Update it, or create a new one.'
    );
    expect(tooltips[0]).toBeInTheDocument();
  });

  it('restores the previous connector if the user leaves the input empty', () => {
    const onChangeHandler = jest.fn();
    render(
      <ConnectorsDropdown {...props} selectedConnector="jira-1" onChange={onChangeHandler} />,
      {
        wrapper: ({ children }) => <TestProviders>{children}</TestProviders>,
      }
    );

    userEvent.click(screen.getByTestId('comboBoxClearButton'));
    act(() => {
      screen.getByTestId('comboBoxSearchInput').blur();
    });

    expect(screen.getByTestId('comboBoxInput')).toHaveTextContent(/Jira/);
    expect(onChangeHandler).toHaveBeenLastCalledWith('jira-1');
  });

  it('shows the add connector button when appendAddConnectorButton is passed', () => {
    render(<ConnectorsDropdown {...props} appendAddConnectorButton={true} />, {
      wrapper: ({ children }) => <TestProviders>{children}</TestProviders>,
    });

    userEvent.click(screen.getByTestId('comboBoxToggleListButton'));

    expect(screen.getByTestId('dropdown-connector-add-connector')).toBeTruthy();
  });
});
