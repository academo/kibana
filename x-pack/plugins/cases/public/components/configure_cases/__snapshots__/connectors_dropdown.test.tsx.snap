// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectorsDropdown formats the connectors correctly 1`] = `
Array [
  Object {
    "data-test-subj": "dropdown-connector-no-connector",
    "label": "No connector selected",
    "value": "none",
  },
  Object {
    "data-test-subj": "dropdown-connector-servicenow-1",
    "label": "My Connector",
    "value": "servicenow-1",
  },
  Object {
    "data-test-subj": "dropdown-connector-resilient-2",
    "label": "My Connector 2",
    "value": "resilient-2",
  },
  Object {
    "data-test-subj": "dropdown-connector-jira-1",
    "label": "Jira",
    "value": "jira-1",
  },
  Object {
    "data-test-subj": "dropdown-connector-servicenow-sir",
    "label": "My Connector SIR",
    "value": "servicenow-sir",
  },
  Object {
    "data-test-subj": "dropdown-connector-servicenow-uses-table-api",
    "label": "My Connector",
    "value": "servicenow-uses-table-api",
  },
]
`;
